typing/jkind.ml

```diff
@@ -2278,32 +2283,207 @@ let for_boxed_variant cstrs =
+
+       {[
+         type 'a t = A : 'b option -> 'b t
+       ]}
+
+       we turn the ['b option] field into ['a option], by substitutiong ['b] for ['a].
+
+       There are a few ways this is slightly subtle ([2a]-[2c]):
+
+       2a. For repeated types on arguments, eg in the following type:
+
+       {[
+         type ('x, 'y) t = A : 'a -> ('a, 'a) t
+       ]}
+
+       we take only the *first* time we see an argument, treating that as an index, and
+       omit all other possible instantiations. That means that in the above type, we'll
+       substitute ['x] for both instances of ['a] and infer a kind of [immutable_data with
+       'x]. This is sound, but somewhat restrictive; in a perfect world, we'd infer a kind
+       of [immutable_data with ('x OR 'y)], but that breaks type inference. At some point
+       in the future, we should at least change the subsumption algorithm to accept either
+       [immutable_data with 'x] or [immutable_data with 'y] (* CR layouts v2.8: do that *)
+
+       2b. If a type appears in an argument as something other than a direct TVar, eg in
+       the following type:
+
+       {[
+         type 'a t : A : 'a -> 'a option t
+       ]}
+
+       we just treat it as if it was existential (see [2c]); this is notably different
+       than what [Datarepr.constructor_existentials] considers to be an existential type
+       variable.
+
+       2c. We replace all existential type variables (including those that appear in the
+       return type, but only nested under some other type), except for row variables, with
+       a [Tof_kind] representing the kind of those existential type variables. Since those
+       have best kinds, they'll just get normalized away during normalization, except in
+       the case that they show up as an argument to a type constructor representing an
+       abstract type - in which case, they still end up in the (fully normalized)
+       with-bounds.
+
+       3. Take the substituted argument types and modalities and add them to the
+       with-bounds for the overall variant.
+    *)
+    let module Variant_parts = struct
+      (** Information about the parts of a variant, used to build up its with-bounds.
+
+          See the Note [With-bounds for GADTs] above *)
+      type t =
+        { (* These first two fields will always have the same length. *)
+          cstr_arg_tys : type_expr list;
+          cstr_arg_modalities : Mode.Modality.Value.Const.t list;
+          params : type_expr list;
+          ret_args : type_expr list;
+          seen : Btype.TypeSet.t
+        }
+
+      let empty =
+        { cstr_arg_tys = [];
+          cstr_arg_modalities = [];
+          params = [];
+          ret_args = [];
+          seen = Btype.TypeSet.empty
+        }
+    end in
+    let { Variant_parts.cstr_arg_tys; cstr_arg_modalities; ret_args; params; _ }
+        =
+      (* This left fold is step 1 from Note [With-bounds for GADTs] *)
```

Comment:
Now I see why the Note is the way it is: the algorithm builds up the information from all constructors into one big pile. I guess that works, according to the invariant Stephen has averred. But I see no need to do this (and doing it is wasteful, making for a harder unification problem in `apply`). Instead, just `apply` once per constructor (well, build up the arguments to `apply` once per constructor, and then actually apply those to each field). This also means that e.g. `seen` is maintained only within one constructor, instead of over the whole type.